Record vectors of stochastic variables when `stochasticOnly = TRUE`

davidcerny opened the pull request:
Fixes #899.

davidcerny committed:
Allow mnModel(..., stochasticOnly=TRUE) to monitor vectors

bjoelle reviewed (commented):
I don't understand why this works as a fix - the `!the_node->isElementVariable()` check is also present when `stochastic_nodes_only` is false, so it should not be the cause of the different behaviour between `stochasticOnly` options

davidcerny commented:
Well, it quickly turned out that it doesn't really "work" – it does make the issue go away, but at the cost of introducing a new one. I'll need to think about the logic of that conditional more carefully.

It's weird that stochastic vector elements are getting captured at all (even when `stochasticOnly = FALSE`), since they don't satisfy either of the two conditions!

davidcerny committed:
It takes much more than that... :-/

davidcerny commented:
So it turns out that when `ModelMonitor::resetDagNodes()` iterates over the model graph, it encounters each vector twice: first as a single node, and then as a set of nodes corresponding to individual vector elements. Since we don't want to include each vector element twice, we keep the former and suppress the latter by requiring `!the_node->isElementVariable()`. However, that's where issue #899 comes from: since a vector of stochastic nodes is itself a deterministic node, we can't extract these nodes to record their values.

This branch goes about it the opposite way: it suppresses the vector-node and iterates over the nodes corresponding to its elements. This change causes multiple test failures, but it seems to me that in most (though not all) cases, it's the new behavior that's correct and the current behavior that's wrong. Here is a detailed description of the differences:

`test_basics/ignore.Rev`

- The `y[1][1]`, `y[1][2]`, `z[1][1][1]`, `z[1][1][2]`, `z[1][2][1]`, `z[1][2][2]` nodes are included in the new output but not the expected output. That's a bug on this branch – I need to do a better job of identifying which nodes are clamped.

`test_Coal_skyline/mcmc_Coal_skyline_fixed.Rev`

- The current (i.e., expected) output only shows 6 significant digits for vector elements, whereas the new output shows 7, which is the default `outputPrecision`. It makes sense for the same precision to be applied to both scalars and vector elements, so I'd argue the new output is correct here.

`test_coalescent/mcmc_heterochronous_GMRF_maptreebased.Rev`

- Again, 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_BSP.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_EBSP.Rev`

- The new output doesn't include the `indicator[1]` column. However, the script shows that while the remaining elements of the `indicator` vector are deterministic variables, `indicator[1]` is constant. So the new behavior is correct. The script produces 2 log files; this accounts for discrepancies in both.

`test_coalescent/mcmc_isochronous_GMRF_maptreebased.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_GMRF_treebased.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_GMRF.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_HSMRF.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_piecewise_3const.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_piecewise_3lin.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_piecewise_6diff.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_Skygrid.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_Skyline.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_isochronous_Skyride.Rev`

- 7 sig. figs instead of 6. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_conversions/UniformInteger.Rev`

- The `y[1]`, ..., `y[100]` nodes are included in the expected output but not the new output. This seems correct to me – these nodes are all clamped, so recording their values is redundant.

`test_EBD/mcmc_EBD.Rev`

- 7 sig. figs instead of 6.

`test_EBD/mcmc_EBD_empirical.Rev`

- 7 sig. figs instead of 6.

`test_HSMRF_BDP/mcmc_HSMRF_BDP.Rev`

- 7 sig. figs instead of 6.

`test_large_model_graph/mcmc_large_normal_model.Rev`

- 7 sig. figs instead of 6.

`test_mixtures/mcmc_mixture.rev`

- The `pi[1][1]`, `pi[1][2]`, ..., `pi[2][4]` nodes are included in the new output but not the expected output. The new output is clearly correct here: the individual elements of `pi` are stochastic nodes in the form of vectors drawn from `dnDirichlet()`. It looks like the current system can handle vectors of i.i.d. stochastic variables (that is, unless we set `stochasticOnly = TRUE`), but not vectors which themselves represent a single draw from a distribution.

`test_mixtures/mcmc_mixture2.rev`

- As above.

`test_Mk/mcmc_Mk.rev`

- 7 sig. figs instead of 6.

`test_Mk/mcmc_Mki.rev`

- 7 sig. figs instead of 6.

`test_Mk/mcmc_Mkv.rev`

- 7 sig. figs instead of 6.

`test_negCladeConst/negativeCladeConstraint.Rev`

- 7 sig. figs instead of 6.

`test_Partition/mcmc_Partition_codon.Rev`

- The `er[1][1]`, `er[1][2]`, ..., `er[4][6]` and `gamma_rates[1][1]`, `gamma_rates[1][2]`, ..., `gamma_rates[4][4]` nodes are included in the new output but not the expected output. This represents another case of vectors that have been created "wholesale" (rather than element by element) being invisible to the current system. The elements of `er` again represent draws from `dnDirichlet()` (i.e., each one is a stochastic node), and the elements of `gamma_rates` were obtained using `fnDiscretizeGamma()` (i.e., each one is a deterministic node). The script produces 2 log files; this accounts for discrepancies in both.

`test_Partition/mcmc_Partition_gene.Rev` 

- The `er[1][1]`, `er[1][2]`, ..., `er[2][6]` and `gamma_rates[1][1]`, `gamma_rates[1][2]`, ..., `gamma_rates[2][4]` nodes are included in the new output but not the expected output. The reason for this is the same as above. The script produces 2 log files; this accounts for discrepancies in both.

`test_Partition/mcmc_Partition_uniform.Rev`

- The `er[1]`, ..., `er[6]` and `gamma_rates[1]`, ..., `gamma_rates[4]` nodes are included in the new output but not the expected output. The reason for this is the same as above. The script produces 2 log files; this accounts for discrepancies in both.

`test_skygrid/mcmc_skygrid.Rev`

- 7 sig. figs instead of 6.

`test_steppingstone/ML_partition_uniform.Rev`

- 7 sig. figs instead of 6. The script produces 10 log files; this accounts for discrepancies in all 10.

`test_steppingstone/ML_partition_uniform_one_stone_at_a_time.Rev`

- 7 sig. figs instead of 6. The script produces 10 log files; this accounts for discrepancies in all 10.

`test_stoch_map/mcmc_corr.Rev`

- This is a very interesting case. There are 24 nodes (called `rates[1][1]`, `rates[1][2]`, ..., `rates[4][4]`) that are included in the new output but not the expected output, but only 16 have unique names – the remaining 8 are duplicates. The reason for this becomes clear once we look at the original script, where we first create a vector of 16 constant nodes, and then replace 8 of its elements by deterministic nodes. This is a bug on this branch. We should be recording just those 8 deterministic nodes.

`test_UCLD_noncentered/mcmc_UCLD_noncentered.rev`

- 7 sig. figs instead of 6.

I'm going to go ahead and change the expected output of those tests for which I believe the new results to be the correct ones.

davidcerny committed:
Replace expected output

davidcerny committed:
Fix for ignoring clamped elements of nested vectors

davidcerny committed:
One more replacement of expected output

davidcerny committed:
Unrelated fix to the help page for dnBeta

davidcerny committed:
Merge branch 'development' into stochastic-only-vectors

davidcerny committed:
Actually, sometimes we do need to monitor the vector itself

rather than its elements. This helps us properly handle

edge cases such as x ~ dnDirichlet([1, 1]); y := x[1].

The expectation here would be to have 3 variables to monitor

(x[1], x[2], y), but if we monitor the elements of the x vector

rather than the vector itself, we will not capture x[2].

davidcerny committed:
Drastic performance improvement for large vectors

davidcerny committed:
Fix the problem w/ missing log file headers in a test

(mvFossilTimeSlideUniform_checkpointing.Rev) that (re)starts

an analysis from a pre-existing checkpoint file but no

pre-existing log files

davidcerny committed:
Preserve insertion order (& thus lexicographic sorting) in JSON

davidcerny commented:
There are a few more failing tests that were either recently added to development, or whose failures were previously masked by the fact that we disabled them for macOS:

`test_BDSTP/BDSTP.Rev`

- The `F[1]`, ..., `F[14]` nodes are included in the expected output but not the new output. This is correct – these nodes are all clamped, so recording their values is redundant. (In fact, they don't even represent any real, interpretable quantities – they are just an artefact of the tortured way in which we currently handle tip age uncertainty.)

`test_coalescent/mcmc_heterochronous_BSP.Rev`

- 7 sig. figs instead of 6 in the `pop_size` columns. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_coalescent/mcmc_heterochronous_GMRF_treebased.Rev`

- 7 sig. figs instead of 6 in the `delta_log_population_size` columns. The script produces 3 log files; this accounts for discrepancies in all 3.

`test_FBD/FBD_dating_ex3.Rev`

- 7 sig. figs instead of 6.

`test_FBD/FBD_one_parameter_skyline.Rev`

- 7 sig. figs instead of 6 in the `psi` columns.

`test_FBD/FBDP_fossil_tip_age_uncertainty.Rev`

- The `F[1]`, ..., `F[14]` nodes are included in the expected output but not the new output. See above.

`test_FBD/FBDP.Rev`

- The `F[1]`, ..., `F[14]` nodes are included in the expected output but not the new output. See above.

`test_FBD/mvFossilTimeSlideUniform_checkpointing.Rev`

- The exact discrepancy here was obscured by issue #911, but when the latter is solved, it's clear the difference is the same as in other BDSTP / FBD tests (see above): the `F[1]`, ..., `F[14]` nodes are included in the expected output but not the new output.

`test_FBD/mvFossilTimeSlideUniform_multiple_runs.Rev`

- The `F[1]`, ..., `F[43]` nodes are included in the expected output but not the new output. See above. The script produces 2 log files; this accounts for discrepancies in both.

`test_MkError/mcmc_MkError.rev`

- 7 sig. figs instead of 6 in the `br_lens` columns.

`test_Partition/mcmc_Partition_codon.Rev`

- Two more differences became apparent when examining the JSON output files. First, instead of recording the branch length vector as follows:

  ```json
  "bl":[0.04668089246107779,0.08518248961710673,0.24508170576439664,...],
  ```
  
  we record it as:
  
  ```json
  "bl[1]":0.04668089246107779,"bl[2]":0.08518248961710673,"bl[3]":0.24508170576439664,...
  ```
  
  This is different, but IMO neither obviously better nor worse.
  
  Second, we also record the nested `er` and `gamma_rates` vectors in the JSON files generated as part of the new output but not of the expected output. This difference was already pointed out for the standard (i.e., tab-separated rather than JSON) output in the comment above. The new output is clearly correct here, since these are stochastic nodes that our current monitoring system ignores for no good reason. Note how these nested vectors are written to JSON:
  
  ```json
  "gamma_rates[1]":[0.8959733745739022,0.9709424352126861,1.0250762799985929,1.1080079102148188],"gamma_rates[2]":[0.8977148201454996,0.9714712971734394,1.0246822805217213,1.1061316021593397],...
  ```
  
  The script produces 2 JSON files; this accounts for discrepancies in both.

`test_proposed_bad_index/model_average_primates_cytb.Rev`

- 7 sig. figs instead of 6.

`test_proposed_bad_index/model_average_primates_cytb_debug.Rev`

- 7 sig. figs instead of 6.

davidcerny committed:
Another replacement of expected output, including for MPI

davidcerny committed:
Repurpose an existing test to further test stochasticOnly=TRUE

davidcerny commented:
This is ready for review, I think. Two final notes:

(1) Performance-wise, the new solution is mostly indistinguishable from the old, but there are exceptions. The [large model graph test](https://github.com/revbayes/revbayes/blob/master/tests/test_large_model_graph/scripts/mcmc_large_normal_model.Rev), which uses `mnModel` to sample a vector of 15,000 stochastic variables 100 times (1 sample every 10 MCMC generations, for a total of 1,000 generations), currently takes about 23 seconds to run on my machine, but on this branch the runtime goes up to about 53 seconds.

(2) When I stated in a previous comment that this branch

> suppresses the vector-node and iterates over the nodes corresponding to its elements

this wasn't completely accurate. An interesting edge case is provided by the [MC3 heat tuning and checkpointing test](https://github.com/revbayes/revbayes/blob/master/tests/test_MC3/scripts/Test_heat_tuning_and_checkpointing.Rev), which contains the following Rev code:

https://github.com/revbayes/revbayes/blob/cbe72b0372cf2c869d951293b383181bacdb9e99/tests/test_MC3/scripts/Test_heat_tuning_and_checkpointing.Rev#L56-L58

Now, if we had _just_ `root_freqs` and no `rtfq1`, the simplistic solution would work just fine. Here, `root_freqs` is a single stochastic node, rather than a vector of two such nodes:

<img width="231" height="234" alt="Screenshot 2025-12-02 at 23 33 24" src="https://github.com/user-attachments/assets/050cfcf4-9efb-4653-a561-ac42cba166fb" />
</br></br>

The notion of suppressing a vector in favor of its elements therefore doesn't apply. The outcome is still as desired – we get two columns in the resulting log file, called `root_freqs[1]` and `root_freqs[2]` – but it was the `Simplex` object itself that handled its own "unfolding" into two elements. It wasn't unfolded by a walk across the model graph, in which `root_freqs` appears as an atom.

If we defined deterministic nodes for _both_ elements of `root_freqs` – i.e., `rtfq1 := root_freqs[1]; rtfq2 := root_freqs[2]` – we wouldn't have an issue either. In doing so, we have quietly created two deterministic nodes called `root_freqs[1]` and `root_freqs[2]`:

<img width="799" height="612" alt="Screenshot 2025-12-02 at 23 43 38" src="https://github.com/user-attachments/assets/614b315b-dfd2-4a7c-9bd8-a4eb230e28f9" />
</br></br>

We correctly suppress `root_freqs` in favor of these two, and the outcome is once again what we'd expect: we have four columns in the resulting log file, labeled `root_freqs[1]`, `root_freqs[2]`, `rtfq1`, and `rtfq2`.

It's only the code from the original test script that poses an issue:

<img width="419" height="612" alt="Screenshot 2025-12-02 at 23 50 40" src="https://github.com/user-attachments/assets/47e3f661-4510-4dfe-9e65-82c783b197f3" />
</br></br>

We have implicitly created a deterministic node called `root_freqs[1]`, which causes us to suppress `root_freqs` – but then we are left with no information about `root_freqs[2]`, and no such column will therefore appear in the output file.

This PR solves the problem by first checking if the number of element-nodes of a given vector-node corresponds to that vector's size. If it does, then it suppress the vector in favor of the elements, as stated in my previous comment. If it doesn't, it instead suppresses the elements in favor of the vector, as we currently do on development.

Is this edge case weird and artificial? Yes. Both before and after this PR we would still record `root_freqs[1]` even if we didn't define a deterministic node for it, so the extra node adds absolutely nothing – just a duplicated column in the log file. But it's probably better to handle edge cases correctly even if they are weird and artificial.

bjoelle commented on src/core/monitors/ModelMonitor.cpp:
```diff
@@ -56,35 +74,101 @@ void ModelMonitor::resetDagNodes( void )
         
+        // Step 3: add variables, including vector elements but not the vectors themselves (unless we made an exception
+        // in the previous step)
+        for (DagNode* the_node : n)
```
I'm wondering if part of the performance hit is because we are running this loop several times
it seems to me we could do this only once, building the list of nodes that should be added as we go and then just adding those
also it seems that the clamped & hidden check could be done way earlier since we don't care about these nodes

davidcerny replied:
@bjoelle In this case, I don't think we can merge the loops. We do need two passes: the first one to collect all vector-nodes and element-nodes in the graph, and the second one to add those we have determined should be added. We can't add nodes as we collect them, because we first need to run Step 2 to decide if we should suppress the vector-node in favor of its element-nodes, or _vice versa_.

In any case, this doesn't really matter. I've actually experimented with lots of ways to make the new code more performant: move some of the checks upfront and add early exits/`continue`s for them, like you suggested; use unordered sets and unordered maps for faster lookup; add non-vector and non-element nodes in Step 1 rather than Step 3; or even cache node names instead of repeatedly calling `the_node->getName()`. None of it made any difference at all, because `ModelMonitor::resetDagNodes()` is not where the bottleneck is. It only gets called twice per analysis, and takes only a few milliseconds each time. `VariableMonitor::monitorVariables()` has a much bigger effect, but still doesn't seem to be the main culprit.

bjoelle replied:
ok, fair enough

bjoelle commented on src/core/monitors/Monitor.cpp:
```diff
@@ -363,7 +363,7 @@ void Monitor::setMcmc(Mcmc *m)
  * Sort the nodes by name so that the order is guaranteed of replicated runs.
  */
-void Monitor::sortNodesByName( void )
+void Monitor::sortNodesByName( bool natural )
```
could be good to mention in a comment what is a "natural" sort as opposed to the default ?

davidcerny replied:
done

bjoelle reviewed (changes requested):
overall this looks good to me
I think the algo can be reorganized to minimize the performance hit however, see detailed comments

davidcerny committed:
Add a comment explaining natural sort
